fvm/fvm_test.go

```diff
@@ -3013,4 +3013,102 @@ func TestEVM(t *testing.T) {
 			))
 		}),
 	)
+
+	// this test makes sure the execution error is correctly handled and returned as a correct type
+	t.Run("execution reverted", newVMTest().
+		withBootstrapProcedureOptions(fvm.WithSetupEVMEnabled(true)).
+		withContextOptions(fvm.WithEVMEnabled(true)).
+		run(func(
+			t *testing.T,
+			vm fvm.VM,
+			chain flow.Chain,
+			ctx fvm.Context,
+			snapshotTree snapshot.SnapshotTree,
+		) {
+			script := fvm.Script([]byte(fmt.Sprintf(`
+				import EVM from %s
+				
+				pub fun main() {
+					let bal = EVM.Balance(flow: 1.0);
+					let acc <- EVM.createBridgedAccount();
+					// withdraw insufficient balance
+					destroy acc.withdraw(balance: bal);
+					destroy acc;
+				}
+			`, chain.ServiceAddress().HexWithPrefix())))
+
+			_, output, err := vm.Run(ctx, script, snapshotTree)
+
+			require.NoError(t, err)
+			require.Error(t, output.Err)
+			require.True(t, errors.IsEVMError(output.Err))
+
+			// make sure error is not treated as internal error by Cadence
+			var internal cadenceErrors.InternalError
+			require.False(t, errors.As(output.Err, &internal))
+		}),
+	)
+
+	// this test makes sure the EVM error is correctly returned as an error and has a correct type
+	// we have implemented a snapshot wrapper to return an error from the EVM
+	t.Run("internal evm error handling", newVMTest().
+		withBootstrapProcedureOptions(fvm.WithSetupEVMEnabled(true)).
+		withContextOptions(fvm.WithEVMEnabled(true)).
+		run(func(
+			t *testing.T,
+			vm fvm.VM,
+			chain flow.Chain,
+			ctx fvm.Context,
+			snapshotTree snapshot.SnapshotTree,
+		) {
+			errWrappers := []evmErrorSnapshot{{
+				snapshotTree: snapshotTree,
+				err:          types.ErrNotImplemented,
+				errChecker:   types.IsAFatalError,
+			}, {
+				snapshotTree: snapshotTree,
+				err:          types.NewStateError(fmt.Errorf("test state error")),
+				errChecker:   types.IsAStateError,
+			}}
+
+			for _, e := range errWrappers {
+				script := fvm.Script([]byte(fmt.Sprintf(`
+					import EVM from %s
+					
+					pub fun main() {
+						destroy <- EVM.createBridgedAccount();
+					}
+				`, chain.ServiceAddress().HexWithPrefix())))
+
+				_, output, err := vm.Run(ctx, script, e)
+
+				require.NoError(t, output.Err)
+				require.Error(t, err)
+				// make sure error it's the right type of error
+				require.True(t, e.errChecker(err))
+			}
+		}),
+	)
+}
+
+// evmErrorSnapshot will return an error once it gets to reading the EVM
+// registers for address allocator, that way we can at that point return a provided
+// error, that error type is then checked against the provider checker.
+type evmErrorSnapshot struct {
+	snapshotTree snapshot.SnapshotTree
+	err          error
+	errChecker   func(error) bool
+}
+
+func (s evmErrorSnapshot) Get(
+	id flow.RegisterID,
+) (
+	flow.RegisterValue,
+	error,
+) {
+	if id.Key == "AddressAllocator" {
+		return nil, s.err
+	}
+
+	return s.snapshotTree.Get(id)
```

Comment:
hmm, I guess there should be a MockedSnapshot somewhere either created by mockery or implemented with func methods, If it doesn't exist we should create it in my view, that way we have more clear tests that are not dependent on a logic outside of the test.

Reply:
@janezpodhostnik do you know if we have one ?

Reply:
I can replace with mockery yeah.

Reply:
I've checked before for existing ones but only found, one in accounts that is just an implementation for tests here https://github.com/onflow/flow-go/blob/f17e073ae9d00e723a5d2e5fafe481a361030c5c/fvm/accounts_test.go#L22 and one in util https://github.com/onflow/flow-go/blob/0596ab06b1225f92b02f539234d292d2a02f619a/cmd/util/ledger/util/util.go#L67

So will generate one with mockery for now.

Reply:
@ramtinms fixed